drivers/timer/arm_arch_timer.c

```diff
@@ -98,6 +98,22 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
 	uint64_t curr_cycle = arm_arch_timer_count();
+
+	/*
+	 * Continuously reading the cntvct_el0 within a short time interval may
+	 * get the same value so that it may cause a side effect when setting
+	 * the next timeout interrupt especially for the board with a
+	 * low-frequency clock.
+	 * For example, On FVP platform, When a timer compare interrupt occurs,
+	 * we got 530000 from cntvct_el0 in the arm_arch_timer_compare_isr
+	 * function and we may also get 530000 from cntvct_el0 here. This will
+	 * cause the req_cycle to round up to the wrong value and cause the
+	 * timer compare interrupt to occur earlier than expected.
+	 */
+	if (curr_cycle == last_cycle) {
+		curr_cycle += 1;
+	}
```

Comment:
this feels a bit like cheating. Can we have this under an `ifdef` on the platform?

Reply:
It's no problem to get this under an `ifdef`. For now, this is related to 4 FVP boards (or at least 2 SoCs, fvp aarch32 and fvp aarch64). 
But AFAIK, some hardware board also has this issue. Then the macro would be too much I think. And this logic won't affect the normal case and even could deal with the issue caused by a slower clock.
Well, it is ok for me to add the macro or not. Do you think it's better to not with the `ifdef`?

Reply:
Or I may miss understanding your `ifdef`? Do you mean a feature macro, then the board with the issue select this?

Reply:
> some hardware board also has this issue

Are there really boards out there as slow as FVP emulation?

> Do you mean a feature macro, then the board with the issue select this?

No, I meant an `ifdef` on the platform `#ifdef CONFIG_FVP_*` or equivalent.

Reply:
Got it.

Reply:
This conversation may out of date. Could I just click `resolve conversation` directly? @carlocaione @povergoing